### PR TITLE
fix endpointslicemirroring controller not create endpointslice when the endpoints are recreate

### LINKED ISSUE #125354: endpointslicemirroring controller not create endpointslice when the endpoints are recreate
### What happened?

* create a service without selector, and manual create endpoints for this service
* endpointslicemirroring controller will create an endpointslice for this service.
* after endpoint slice created, and kube-controller-manager restart.
* during the kube-controller-manager restart, the endpoint was deleted and recreated
* after kube-controller-manager restart,  garbagecollector controller delete this endpointslice and endpointslice not recreate until endpoint/service update event or kube-controller-manager restart

### What did you expect to happen?

endpointslice will be recreated

### How can we reproduce it (as minimally and precisely as possible)?

* create a service without selector, and manual create endpoints for this service
* endpointslicemirroring controller will create an endpointslice for this service.
* after endpoint slice created, and kube-controller-manager restart.
* during the kube-controller-manager restart, the endpoint was deleted and recreated

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
v 1.30.0
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
kind 1.30.0
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by yangjunmyfm192085**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When kube-controller-manager is restarted, the endpointSlice information in endpointSliceTracker is empty. At this time, if an endpointSlice add event is received, it is updated to endpointSliceTracker.
#### Which issue(s) this PR fixes:

Fixes #125354

#### Does this PR introduce a user-facing change?

```release-note
Fix: during the kube-controller-manager restart, when the corresponding Endpoints resource was manually deleted and recreated, causing the endpointslice to fail to be created normally.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by yangjunmyfm192085 — COMMENTED** [review]
/retest

**Comment by aojea** [comment]
can we have an unit test reproducing this problem?

**Comment by yangjunmyfm192085** [comment]
> can we have an unit test reproducing this problem?

sure, I have updated the unit test to simulate the scenario where `kube-controller-manager` starts up, loads `endpointSlice`, and then `endpointSlice` is deleted.

**Comment by aojea** [comment]
> loads `endpointSlice`, and then `endpointSlice` is deleted.

who deletes the endpointSlice?

/cc @robscott 

you may want to look at this issue, see also the reference, I think kind of understand the sequence of events that causes this but I'm not fully sure why the solution is to track the endpointslice in the event handler, why does not reconcile?

**Comment by robscott** [comment]
@aojea thanks for the ping on this! I left a [comment on the original issue](https://github.com/kubernetes/kubernetes/issues/125354#issuecomment-2163969836) explaining how I understand this bug and corresponding fix.

@yangjunmyfm192085 thanks for finding and fixing this bug! I'll approve, but will defer to someone else to LGTM since this is a fairly complex one and would like another set of eyes just to be sure.

/approve
/cc @gauravkghildiyal @swetharepakula

**Comment by robscott** [comment]
@yangjunmyfm192085 Can you please add a release note in your PR description that describes the bug fix?

/hold

**Comment by yangjunmyfm192085** [comment]
> @yangjunmyfm192085 Can you please add a release note in your PR description that describes the bug fix?
> 
> /hold

yeah, I have updated

**Comment by robscott** [comment]
>Fix: When kube-controller-manager is restarted, the endpointslice created by the endpointslice mirroring controller is deleted and rebuilt, causing the endpointslice to fail to be created normally.

@yangjunmyfm192085 I thought the problem only happened when the corresponding Endpoints resource was manually deleted?

**Comment by yangjunmyfm192085** [comment]
> > Fix: When kube-controller-manager is restarted, the endpointslice created by the endpointslice mirroring controller is deleted and rebuilt, causing the endpointslice to fail to be created normally.
> 
> @yangjunmyfm192085 I thought the problem only happened when the corresponding Endpoints resource was manually deleted?

May be deleted by the garbagecollector controller
refer:https://github.com/kubernetes/kubernetes/issues/125354

**Comment by robscott** [comment]
>May be deleted by the garbagecollector controller

Right, but the garbagecollector controller would only delete the EndpointSlice if the owner no longer existed, and the ownerRef of a mirrored EndpointSlice points to the original Endpoints resource. (Also left a more detailed comment on https://github.com/kubernetes/kubernetes/issues/125354#issuecomment-2163969836).

**Comment by gauravkghildiyal** [comment]
HI @yangjunmyfm192085, thanks for working on this!

Simply looking at the PR and the description, its non obvious to me about what you expect to happen by adding the endpointSlice to the tracker on seeing an onAdd event. While it does make logical sense to me that the endpointSlice tracker should get accurately rebuilt after restart, I'm trying to reason out why this problem doesn't manifest in some more obvious ways.

What would be helpful is a few more details about the proposed solution. I'd like to ensure that the solution we are making is the right one, and that we don't need to fix this in a different.

**Comment by gauravkghildiyal** [comment]
Thanks! That explains the problem. What at least I'd need help with is details about the proposed solution.

**Comment by songxiao-wang87** [comment]
After testing, modification can solve this problem.

**Comment by aojea** [comment]
@robscott @yangjunmyfm192085 I still don't have clear where this Add event on endpointslice is coming

Is the old endpointslice that is going to be deleted but the informer is restarting and still has it?

**Comment by yangjunmyfm192085** [comment]
> @robscott @yangjunmyfm192085 I still don't have clear where this Add event on endpointslice is coming
> 
> Is the old endpointslice that is going to be deleted but the informer is restarting and still has it?

@aojea Add event on endpointslice triggered afte kube-controller-manager process restart

**Comment by aojea** [comment]
> @aojea Add event on endpointslice triggered afte kube-controller-manager process restart

then, this has to be the slice that later will be deleted by the garbage collector, no?

**Comment by yangjunmyfm192085** [comment]
> then, this has to be the slice that later will be deleted by the garbage collector, no?
Yes, it was indeed deleted by the garbage collector

**Comment by yangjunmyfm192085** [comment]
> can we have a timeline of events so we can understand better the scenario? is it something like?
> 
> t0 - kube-controller-manager stop t1 - endpoint is deleted t2 - endpoint is created t3 - kube-controller manager starts t4 - mirroring controller does not do anything because sees the endpoint and the endpointslice t5 - endpointslice is deleted by the garbage collector
> 
> if this is the case this solve the problem as a side effect, IMHO the reconciler should detect that the reference on the mirrored slice is not matching the current endpoint (UUID should be different) and reconcile it
> 
> /lgtm cancel


Yes, That's the issue as you described.

**Comment by aojea** [comment]
Then two questions:
1. Why the reconciler on the addition of the endpoint does not detect the existing slice is stale and not referencing the new endpoint (different uid) ?
2. why the endpointSliceDelete hook does not reconcile?

**Comment by yangjunmyfm192085** [comment]
> * Why the reconciler on the addition of the endpoint does not detect the existing slice is stale and not referencing the new endpoint (different uid) ?
> * why the endpointSliceDelete hook does not reconcile?



> Then two questions:
> 
> 1. Why the reconciler on the addition of the endpoint does not detect the existing slice is stale and not referencing the new endpoint (different uid) ?
> 2. why the endpointSliceDelete hook does not reconcile?



1. The current reconcile process does not check the uid of the endpoints corresponding to the endpointslice ownerReferences
2. When the endpointslice is deleted, does not reconcile because no record is found in the endpointSliceTracker. This is what the pr wants to fix

**Comment by aojea** [comment]
> When the endpointslice is deleted, does not reconcile because no record is found in the endpointSliceTracker. This is what the pr wants to fix

yeah, but we are adding a endpointslice that we know does not belong to the existing endpoint, aren't we?

> 1. The current reconcile process does not check the uid of the endpoints corresponding to the endpointslice ownerReferences

the bug seems to be this, so we should fix that ...

**Comment by yangjunmyfm192085** [comment]
> > When the endpointslice is deleted, does not reconcile because no record is found in the endpointSliceTracker. This is what the pr wants to fix
> 
> yeah, but we are adding a endpointslice to the tracked that we know does not belong to the existing endpoint, aren't we?
> 
> > 1. The current reconcile process does not check the uid of the endpoints corresponding to the endpointslice ownerReferences
> 
> the bug seems to be this, so we should fix that ...



> > When the endpointslice is deleted, does not reconcile because no record is found in the endpointSliceTracker. This is what the pr wants to fix
> 
> yeah, but we are adding a endpointslice to the tracked that we know does not belong to the existing endpoint, aren't we?
> 
> > 1. The current reconcile process does not check the uid of the endpoints corresponding to the endpointslice ownerReferences
> 
> the bug seems to be this, so we should fix that ...



> should



> > When the endpointslice is deleted, does not reconcile because no record is found in the endpointSliceTracker. This is what the pr wants to fix
> 
> yeah, but we are adding a endpointslice to the tracked that we know does not belong to the existing endpoint, aren't we?
> 
> > 1. The current reconcile process does not check the uid of the endpoints corresponding to the endpointslice ownerReferences
> 
> the bug seems to be this, so we should fix that ...

Yeah, Let me try to fix it.

**Review by yangjunmyfm192085 — COMMENTED** [review]
/cc @aojea  @robscott

**Comment by 249043822** [comment]
I think both the fixes make sense to me. (#^.^#)
The first makes the endpointSliceTracker fully mange the endpointslices lifecycle, so at the start time of the kube-controller-manager, we should initialize all the endpointslices to endpointSliceTracker even throgh it's stale, the reconcile and the event trigger will assure the eventually consistent.
The second one reconciles the endpoint and endpointslice by adding OwnerReference check, this can update the endpintslice correctly, althrogh it maybe fail while the garbagecollector deletes it first, the retry can cover it.

**Comment by aojea** [comment]
> The first makes the endpointSliceTracker fully mange the endpointslices lifecycle, so at the start time of the kube-controller-manager, we should initialize all the endpointslices to endpointSliceTracker even throgh it's stale, the reconcile and the event trigger will assure the eventually consistent.

what if the slice is stale and the endpoint does not longer exist? do we keep it in the tracker?

File: pkg/controller/endpointslicemirroring/reconciler.go [hunk 1]
```diff
@@ -207,7 +207,8 @@ func (r *reconciler) reconcileByPortMapping(
 		totals = totalChanges(existingSlices[0], desiredSet)
 		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
 			apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
-			apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) {
+			apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) &&
+			!needRebuildExistslices(endpoints, existingSlices[0]) {
```
Comment (aojea) [inline pkg/controller/endpointslicemirroring/reconciler.go:211]: unit test is missing, :)

**Review by robscott — COMMENTED** [review]
Thanks @yangjunmyfm192085! This seems like a very reasonable approach that should limit the blast radius of changes.

**Comment by 249043822** [comment]
> > The first makes the endpointSliceTracker fully mange the endpointslices lifecycle, so at the start time of the kube-controller-manager, we should initialize all the endpointslices to endpointSliceTracker even throgh it's stale, the reconcile and the event trigger will assure the eventually consistent.
> 
> what if the slice is stale and the endpoint does not longer exist? do we keep it in the tracker?

if the endpoint is missing,  the garbagecollector will also delete the endpointslice eventually, then the REMOVE Event handler could clear the endpointSliceTracker , so I think it's also a reasonable fix, it deos not cause memory leak

File: pkg/controller/endpointslicemirroring/reconciler.go [hunk 2]
```diff
@@ -334,3 +335,13 @@ func totalChanges(existingSlice *discovery.EndpointSlice, desiredSet endpointsli
 	totals.added = desiredSet.Len() - existingMatches
 	return totals
 }
+
+func needRebuildExistslices(endpoints *corev1.Endpoints, existingSlice *discovery.EndpointSlice) bool {
```
[the remark this answers is not in the document]
Comment (yangjunmyfm192085) [reply pkg/controller/endpointslicemirroring/reconciler.go]: I have updated

**Comment by aojea** [comment]
> if the endpoint is missing, the garbagecollector will also delete the endpointslice eventually, then the REMOVE Event handler could clear the endpointSliceTracker , so I think it's also a reasonable fix, it deos not cause memory leak

let's address the bug that we know, I think technically is correct but the source of truth must be the endpoint, this controller and endpointslices have a lot of edge cases so better if we are cautious here.

/test pull-kubernetes-integration 

unrelated failure

/lgtm

Thanks for the great detail and work on this complex bug

@robscott please check for final approval

**Comment by gauravkghildiyal** [comment]
tl;dr: The solution as it presently is LGTM. Though I still have a few thoughts to share to make sure we are fine with what we get out of this.

---

A very general solution to the problem above would be to say that "if EndpointSlice is deleted, our controller should be smart enough to recognize that the EndpointSlice was still needed and recreate it immediately" (without waiting for the next update to the corresponding Endpoint)

- The _current_ proposed solution ensures that if an EndpointSlice has an incorrect ownerReference, the mirroring controller will update this ownerReference to the correct one, **thereby preventing the deletion of the EndpointSlice by the garbage collector**.

- While the above handles the major issue, it will still fails to handle the general problem. Elaborating on that, if I were to simply:
    1. Restart kube-controller-manager, and
    2. Delete the EndpointSlice.
  (and do nothing else besides these two things). 
   then I would again observe the issue where the EndpointSlice remains deleted until the next update to the Endpoint triggers it's recreation.

The heart of the problem lies in the fact that the `endpointSliceTracker` is not rebuilt to the exact same state that it had before the kube-controller-manager restarted.
- As of now, EndpointSlices are only added to the `endpointSliceTracker` when an EndpointSlice is newly CREATED or UPDATED.
- In other words, when kube-controller-manager is restarted, because the EndpointSlices already existed, they are not re-added to the `endpointSliceTracker`. 
- This inconsistency of the `endpointSliceTracker` after the kube-controller-manager restarts leads to the chain of events where the deletion of the EndpointSlice is not immediately followed by it's automatic recreation.

**The previous proposed solution** solved the general problem by ensuring that when the kube-controller-manager restarts, and the `onEndpointSliceAdd` events are re-played by the informer, they will result in the `endpointSliceTracker` gettting re-populated correctly. **Though I myself felt the same thing that perhaps the solution may have been put in a place which conveyed a different meaning**

Would something like this perhaps be better overall and cover everything?

```diff
  if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
    apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
    apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) &&
+   !needRebuildExistingSlices(endpoints, existingSlices[0]) { // Update EndpointSlices if required.
+  
+   if !r.endpointSliceTracker.Has(existingSlices[0]) {
+     r.endpointSliceTracker.Update(existingSlices[0]) // Always ensure each EndpointSlice is being tracked.
+   }
  
    return slices, totals
  }
```

For posterity, would simply doing the following also have solved both problems? (The difference from the previous proposed solution being that this feels somewhat more consistent than updating the `endpointSliceTracker` within `onEndpointSliceAdd`, and in a way still keeps the source of truth as the Endpoints resource)

```diff
  if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
    apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
    apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) && {

+   // If we plan on returning early, atleast ensure that the endpointSliceTracker
+   // has this EndpointSlice, which is neither being created, nor updated.
+   
+   if !r.endpointSliceTracker.Has(existingSlices[0]) {
+     r.endpointSliceTracker.Update(existingSlices[0])
+   }
  
    return slices, totals
  }
```

**Comment by yangjunmyfm192085** [comment]
> tl;dr: The solution as it presently is LGTM. Though I still have a few thoughts to share to make sure we are fine with what we get out of this.
> 
> A very general solution to the problem above would be to say that "if EndpointSlice is deleted, our controller should be smart enough to recognize that the EndpointSlice was still needed and recreate it immediately" (without waiting for the next update to the corresponding Endpoint)
> 
> * The _current_ proposed solution ensures that if an EndpointSlice has an incorrect ownerReference, the mirroring controller will update this ownerReference to the correct one, **thereby preventing the deletion of the EndpointSlice by the garbage collector**.
> * While the above handles the major issue, it will still fails to handle the general problem. Elaborating on that, if I were to simply:
>   
>   1. Restart kube-controller-manager, and
>   2. Delete the EndpointSlice.
>      (and do nothing else besides these two things).
>      then I would again observe the issue where the EndpointSlice remains deleted until the next update to the Endpoint triggers it's recreation.
> 
> The heart of the problem lies in the fact that the `endpointSliceTracker` is not rebuilt to the exact same state that it had before the kube-controller-manager restarted.
> 
> * As of now, EndpointSlices are only added to the `endpointSliceTracker` when an EndpointSlice is newly CREATED or UPDATED.
> * In other words, when kube-controller-manager is restarted, because the EndpointSlices already existed, they are not re-added to the `endpointSliceTracker`.
> * This inconsistency of the `endpointSliceTracker` after the kube-controller-manager restarts leads to the chain of events where the deletion of the EndpointSlice is not immediately followed by it's automatic recreation.
> 
> **The previous proposed solution** solved the general problem by ensuring that when the kube-controller-manager restarts, and the `onEndpointSliceAdd` events are re-played by the informer, they will result in the `endpointSliceTracker` gettting re-populated correctly. **Though I myself felt the same thing that perhaps the solution may have been put in a place which conveyed a different meaning**
> 
> Would something like this perhaps be better overall and cover everything?
> 
> ```diff
>   if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
>     apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
>     apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) &&
> +   !needRebuildExistingSlices(endpoints, existingSlices[0]) { // Update EndpointSlices if required.
> +  
> +   if !r.endpointSliceTracker.Has(existingSlices[0]) {
> +     r.endpointSliceTracker.Update(existingSlices[0]) // Always ensure each EndpointSlice is being tracked.
> +   }
>   
>     return slices, totals
>   }
> ```
> 
> For posterity, would simply doing the following also have solved both problems? (The difference from the earlier solution [7b93e6e](https://github.com/kubernetes/kubernetes/commit/7b93e6e700cd7e53297be8e3c2b47b3f4ad1f388) proposed in this PR being that this feels somewhat more consistent than updating the `endpointSliceTracker` within `onEndpointSliceAdd`, and in a way still keeps the source of truth as the Endpoints resource)
> 
> ```diff
>   if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
>     apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
>     apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) && {
> 
> +   // If we plan on returning early, atleast ensure that the endpointSliceTracker
> +   // has this EndpointSlice, which is neither being created, nor updated.
> +   
> +   if !r.endpointSliceTracker.Has(existingSlices[0]) {
> +     r.endpointSliceTracker.Update(existingSlices[0])
> +   }
>   
>     return slices, totals
>   }
> ```



> ```diff
> +   if !r.endpointSliceTracker.Has(existingSlices[0]) {
> +     r.endpointSliceTracker.Update(existingSlices[0]) // Always ensure each EndpointSlice is being tracked.
> +   }
> ```



 Just as you said. If endpointslice does not change, endpointSliceTracker also needs to be updated.
I have updated, please PTAL again, thanks

**Comment by aojea** [comment]
@249043822 do you agree?

**Comment by 249043822** [comment]
> @249043822 do you agree?

Yes, LGTM

**Comment by aojea** [comment]
/lgtm
/approve

Thanks
